### DISCUSSION TITLE
Upgrading from 0.19.4 not working

### DISCUSSION THREAD

**Discussion by jramke** [pocketbase/pocketbase#3981] [Q&A] [answered]
Hey i tried to upgrade my pocketbase from 0.19.4 to 0.20.0/1 but when i run `./pocketbase serve` i get this error:
```
Failed to apply migration logs_generalization.go: database disk image is malformed (11)
```

When i clear the `pb_data` directory it works but then i cant get the data back.
The live instance is hosted on pockethost where i get the same error.

I already upgraded pocketbase in another project where everything worked.

Its just a simple app with the user collection and a collection where the users can track their weight. 
Let me know about some informations i could provide for more details

**Comment by ganigeorgiev** [comment]
Do you mean that the error is reproduced even when upgrading locally? On what OS you are testing? Are you testing within a docker environment? If yes, how the volume is mounted?

In any case, I don't think this is caused by the PocketBase upgrade.

Most likely the error is either because of a file permission or maybe you've deleted the db file while the application was still running, or another process was still having an open connection to it (node, litestream, etc.), I'm not really sure.

You can try to run the integrity checks on the db file using the sqlite3 cli tool and check if it will return an error:

```sh
# for the data.db
sqlite3 ./pb_data/data.db "PRAGMA integrity_check"

# for the logs.db
sqlite3 ./pb_data/logs.db "PRAGMA integrity_check"
```

Another thing you can try, since the error mentions the `logs_generalization` migration, is to delete the `pb_data/logs.db` file (and its sidecar -wal and -shm files) and restart the application.

**Reply by ganigeorgiev** [reply]
Also how did you upgrade? Could you provide the exact steps and commands you've executed?

**Reply by ganigeorgiev** [reply]
You can check also ["How To Corrupt An SQLite Database File"](https://www.sqlite.org/howtocorrupt.html), maybe it will help you finding what could have caused the "malformed" error in the first place.

**Comment by jramke** [comment]
Hey, thanks for the quick answers. 

The integrity checks were ok. 
I'm on windows.

I continued with trial and error and made it work with deleting everything from the `pb_data` folder then put the `data.db` from the live instance inside.

Then i got this error: 
```
Failed to apply migration created_weights.js: constraint failed: UNIQUE constraint failed: _collections.name (2067)
```
So i deleted the `pb_migrations/created_weights.js` file and now it works.

This was the content: 
```js
/// <reference path="../pb_data/types.d.ts" />
migrate((db) => {
  const collection = new Collection({
    "id": "mayrchqo4cobnhz",
    "created": "2023-12-17 12:36:34.659Z",
    "updated": "2023-12-17 12:36:34.659Z",
    "name": "weights",
    "type": "base",
    "system": false,
    "schema": [
      {
        "system": false,
        "id": "xgjxzxmh",
        "name": "weight",
        "type": "number",
        "required": true,
        "presentable": false,
        "unique": false,
        "options": {
          "min": null,
          "max": 500,
          "noDecimal": false
        }
      },
      {
        "system": false,
        "id": "lpqrw6lo",
        "name": "kcal",
        "type": "text",
        "required": false,
        "presentable": false,
        "unique": false,
        "options": {
          "min": null,
          "max": null,
          "pattern": ""
        }
      },
      {
        "system": false,
        "id": "ja3xusiy",
        "name": "date",
        "type": "date",
        "required": true,
        "presentable": false,
        "unique": false,
        "options": {
          "min": "",
          "max": ""
        }
      },
      {
        "system": false,
        "id": "4h5h0ntl",
        "name": "user",
        "type": "relation",
        "required": true,
        "presentable": false,
        "unique": false,
        "options": {
          "collectionId": "_pb_users_auth_",
          "cascadeDelete": true,
          "minSelect": null,
          "maxSelect": 1,
          "displayFields": null
        }
      }
    ],
    "indexes": [],
    "listRule": "",
    "viewRule": "",
    "createRule": "@request.auth.id != \"\" && @request.auth.id = @request.data.user",
    "updateRule": "@request.auth.id = user",
    "deleteRule": "@request.auth.id = user",
    "options": {}
  });

  return Dao(db).saveCollection(collection);
}, (db) => {
  const dao = new Dao(db);
  const collection = dao.findCollectionByNameOrId("mayrchqo4cobnhz");

  return dao.deleteCollection(collection);
})
```

Is there a way i can regenerate or fix this file manually?

Best regards

**Reply by ganigeorgiev** [reply, marked as answer]
>  deleting everything from the pb_data folder then put the data.db from the live instance inside.

If you didn't copy also the `data.db-wal` and `data.db-shm` (if they were existing) then this is the reason why the database could have been corrupted.

The above error is not because the migration file is incorrect. What it means is that your database already has the collection existing but it is missing the migration entry with the file name in your `_migrations` table and it tries to create it again. How you ended up in this state, I don't know and there are not enough details to try to guess.

One way to "fix" it is to add the file name manually into the table, but at this point I don't really understand what steps you've taken to end up in the current state, so I'll stop engage further in the discussion to avoid causing further confusion.

Corrupting the database is not something that should be dismissed lightly.
My recommendation is to try to trace back your steps to try to find what has caused it and if you think that it is a PocketBase issue, please create a **minimal** reproducible repo and I'll have a more detailed look.

**Reply by jramke** [reply]
Thanks for the answer im gonna try this.
